### ISSUE TITLE
Reuse enum definition multiple times

### ISSUE THREAD

**Issue by OliverNET** [Azure/azure-functions-openapi-extension#153]
Enum in generated OpenAPI document is defined inline.

```"components": {
  "schemas": {
    "myRequest": {
(...)
      "properties": {
        "myEnumProperty": {
          "enum": [
            "value1",
            "value2"
          ],
          "type": "string",
(...)
```

When using an enum multiple times within the same API, Visual Studio 'Connected Services' (utilizing nswag) will generate multiple enum definitions on client-side for the same enum. This requires uneccessary type conversions in client code.

OpenAPI spec allows to define enum as reference, which results in only one enum definition on client-side:

```"components": {
  "schemas": {
    "myEnumType": {
      "enum": [
        "value1",
        "value2"
      ],
      "type": "string",
(...)
    },
    "myRequest": {
(...)
      "properties": {
        "myEnumProperty": {
          "$ref": "#/components/schemas/myEnumType"
        }
(...)
```

Is it possible to configure enum gerenation as reference?

Would be a great help.

**Comment by changbokLee** [comment]
Hello, I'm not sure if it's the answer, but there's a way to reuse it, so I'm attaching a link. https://swagger.io/docs/specification/data-models/enums/

**Comment by justinyoo** [comment]
> Hello, I'm not sure if it's the answer, but there's a way to reuse it, so I'm attaching a link. https://swagger.io/docs/specification/data-models/enums/

Thanks for the info! Let's find a way to reuse the enum through this extension.

**Comment by justinyoo** [comment]
@changbokLee Would you mind take a look at this issue?

**Comment by changbokLee** [comment]
okay sir  i will try , our project are supporting that swagger version 2.0 3.0? , documention that it is explainning only 2.0 3.0 version .

**Comment by justinyoo** [comment]
@changbokLee Our extension currently relies on [https://github.com/microsoft/OpenAPI.NET](https://github.com/microsoft/OpenAPI.NET) that supports both OpenAPI v2 and v3.0.1. But your codebase should focus on v3.0.1. The rest will be automatically done for you.

**Comment by HermannGruber** [comment]
I guess I'm facing the same issue. I have an azure function, for which the generated swagger.json looks similar as above, but I want to have the enums defined once and then referenced. Currently it reads
```
{
            "name": "pruefungszeitpunkt",
            "in": "query",
            "description": "Prüfungszeitpunkt",
            "schema": {
              "enum": [
                "beiKodierung",
                "zurAbrechnung",
                "waehlbar"
              ],
              "type": "string",
              "default": "beiKodierung"
            },
            "x-ms-visibility": "important"
          }
```

But I would like to see

```
          {
            "name": "pruefungszeitpunkt",
            "in": "query",
            "description": "Prüfungszeitpunkt",
            "schema": {
              "$ref": "#/components/schemas/Pruefungszeitpunkt",
              "type": "string"
            },
            "x-ms-visibility": "important"
          },
```
Can you please tell me whether this will be resolved using the current issue? If not, I will happily open a separate issue. (My generated swagger.json says at the top of the file that it is targeting OpenApi 3.0.1.)

**Comment by HermannGruber** [comment]
Unfortunately I was not able to follow up with the latest discussion on this. Can you confirm that my comment above refers to the same issue? Otherwise, I'd happily open a separate issue.

**Comment by level120** [comment]
I think this issue can't start before solving https://github.com/Azure/azure-functions-openapi-extension/issues/248
Because enum type may have many definitions having different namespaces.

```cs
namespace Some.Alpha
{
    public enum Status { ... }
}

namespace Any.Alpha
{
    public enum Status { ... }
}
```

**Comment by AllTaken** [comment]
@level120 Having this work as long as enum names are unique would already add a lot of value...

**Comment by josh-burton** [comment]
Has there been any progress here?

**Comment by kamilpaszkowski** [comment]
its already so many months and there isnt anything new about it ?

**Comment by Dennis-Rosenbaum** [comment]
What is the status on this, it currently gives us lots of hassle when the enums aren't placed in the schema.

**Comment by RoelVerhees** [comment]
Bump, fix this pls :)

**Comment by justinyoo** [comment]
#610 has addressed this issue. The next `v2-preview` release will contain this fix.

**Comment by MatthijsRademaker** [comment]
@justinyoo When is the next preview scheduled? This is not in v2.0.0-preview2 according to https://github.com/Azure/azure-functions-openapi-extension/compare/v2.0.0-preview1...v2.0.0-preview2

**Comment by L3mur1** [comment]
Hi, was this fixed in 2.0.0 preview2?
i added to local settings:
```
  "Values": {
    "OpenApi__UseFullNamespace": "true",
```
the flag didnt work and open api schemas are still generated as not reusable:
```
          "language": {
            "enum": [
              "None",
              "EN",
              "PL",
              "TT"
            ],
            "type": "string",
            "default": "None"
          }
```
```
    [JsonConverter(typeof(StringEnumConverter))]
    public enum Language
    {
        None,
        EN,
        PL,
        TT = 999,
    }
```

`[OpenApiParameter(name: JokesDocumentation.QueryParameters.Language, In = ParameterLocation.Query, Required = true, Type = typeof(Language), Description = JokesDocumentation.QueryParametersDescriptions.LanguageDescription)]`

**Comment by L3mur1** [comment]
Hi, was this fixed in 2.0.0 preview2?
i updated package to:

`PackageReference Include="Microsoft.Azure.Functions.Worker.Extensions.OpenApi" Version="2.0.0-preview2" `

i added to local settings:
```
  "Values": {
    "OpenApi__UseFullNamespace": "true",
```
the flag didnt work and open api schemas are still generated as not reusable:
```
          "language": {
            "enum": [
              "None",
              "EN",
              "PL",
              "TT"
            ],
            "type": "string",
            "default": "None"
          }
```
```
    [JsonConverter(typeof(StringEnumConverter))]
    public enum Language
    {
        None,
        EN,
        PL,
        TT = 999,
    }
```

`[OpenApiParameter(name: JokesDocumentation.QueryParameters.Language, In = ParameterLocation.Query, Required = true, Type = typeof(Language), Description = JokesDocumentation.QueryParametersDescriptions.LanguageDescription)]`

worth to note im generating:
{
  "openapi": "3.0.1",

am I doing something wrong?
